File: ios/Packages/HapiKit/Sources/HapiProtocol/Models/Machine.swift
import Foundation

/// Machine (runner host) metadata written by the CLI.
///
/// Mirrors `MachineMetadataSchema` (`shared/src/schemas.ts`).
public struct MachineMetadata: Codable, Equatable, Sendable {
    public var host: String
    public var platform: String
    public var happyCliVersion: String
    public var displayName: String?
    public var homeDir: String?
    public var happyHomeDir: String?
    public var happyLibDir: String?
    public var workspaceRoots: [String]?
    /// Machine-scoped RPC capability ids this runner registers.
    public var capabilities: [String]?
    public var startedCliMtimeMs: Double?
    public var installedCliMtimeMs: Double?
    public var supervisedRestart: Bool?

    public init(
        host: String,
        platform: String,
        happyCliVersion: String,
        displayName: String? = nil,
        homeDir: String? = nil,
        happyHomeDir: String? = nil,
        happyLibDir: String? = nil,
        workspaceRoots: [String]? = nil,
        capabilities: [String]? = nil,
        startedCliMtimeMs: Double? = nil,
        installedCliMtimeMs: Double? = nil,
        supervisedRestart: Bool? = nil
    ) {
        self.host = host
        self.platform = platform
        self.happyCliVersion = happyCliVersion
        self.displayName = displayName
        self.homeDir = homeDir
        self.happyHomeDir = happyHomeDir
        self.happyLibDir = happyLibDir
        self.workspaceRoots = workspaceRoots
        self.capabilities = capabilities
        self.startedCliMtimeMs = startedCliMtimeMs
        self.installedCliMtimeMs = installedCliMtimeMs
        self.supervisedRestart = supervisedRestart
    }
}

/// Runner capability flags inside `RunnerState`.
///
/// `agentConfigs` (unified agent configuration descriptors) is kept
/// wire-verbatim; nothing in v1 consumes its internals.
public struct RunnerCapabilities: Codable, Equatable, Sendable {
    public var piExistingSessionResume: Bool?
    public var agentConfigs: [JSONValue]?

    public init(piExistingSessionResume: Bool? = nil, agentConfigs: [JSONValue]? = nil) {
        self.piExistingSessionResume = piExistingSessionResume
        self.agentConfigs = agentConfigs
    }
}

/// Runner daemon state reported per machine.
///
/// Mirrors `RunnerStateSchema` (`shared/src/schemas.ts`). `status` and
/// `shutdownSource` are open string unions on the wire (`enum | string`), so
/// they stay `String` with known-value helpers; `lastSpawnError` is kept
/// wire-verbatim.
public struct RunnerState: Codable, Equatable, Sendable {
    public var status: String
    public var pid: Int?
    public var httpPort: Int?
    public var startedAt: Int?
    public var capabilities: RunnerCapabilities?
    public var shutdownRequestedAt: Int?
    public var shutdownSource: String?
    public var lastSpawnError: JSONValue?

    public var isRunning: Bool { status == "running" }
    public var isShuttingDown: Bool { status == "shutting-down" }

    public init(
        status: String,
        pid: Int? = nil,
        httpPort: Int? = nil,
        startedAt: Int? = nil,
        capabilities: RunnerCapabilities? = nil,
        shutdownRequestedAt: Int? = nil,
        shutdownSource: String? = nil,
        lastSpawnError: JSONValue? = nil
    ) {
        self.status = status
        self.pid = pid
        self.httpPort = httpPort
        self.startedAt = startedAt
        self.capabilities = capabilities
        self.shutdownRequestedAt = shutdownRequestedAt
        self.shutdownSource = shutdownSource
        self.lastSpawnError = lastSpawnError
    }
}

/// Point-in-time host health sample.
///
/// Mirrors `MachineHealthSchema` (`shared/src/schemas.ts`). zod marks it
/// `.strict()`, but this decoder stays lenient on purpose: health is
/// advisory, and a new hub-side field must not break machine decoding.
public struct MachineHealth: Codable, Equatable, Sendable {
    public var collectedAt: Int
    public var cpuCount: Int?
    public var load1m: Double?
    public var cpuPercent: Double?
    public var memoryPercent: Double?
    public var uptimeSeconds: Double?

    public init(
        collectedAt: Int,
        cpuCount: Int? = nil,
        load1m: Double? = nil,
        cpuPercent: Double? = nil,
        memoryPercent: Double? = nil,
        uptimeSeconds: Double? = nil
    ) {
        self.collectedAt = collectedAt
        self.cpuCount = cpuCount
        self.load1m = load1m
        self.cpuPercent = cpuPercent
        self.memoryPercent = memoryPercent
        self.uptimeSeconds = uptimeSeconds
    }
}

/// One machine row of `GET /api/machines` and full-payload `machine-updated`
/// events.
///
/// Mirrors `MachineSchema` (`shared/src/schemas.ts:492-507`).
public struct Machine: Codable, Equatable, Sendable {
    public var id: String
    public var namespace: String
    public var seq: Int
    public var createdAt: Int
    public var updatedAt: Int
    public var active: Bool
    public var activeAt: Int
    public var metadata: MachineMetadata?
    public var metadataVersion: Int
    public var runnerState: RunnerState?
    public var runnerStateVersion: Int
    public var health: MachineHealth?

    public init(
        id: String,
        namespace: String,
        seq: Int,
        createdAt: Int,
        updatedAt: Int,
        active: Bool,
        activeAt: Int,
        metadata: MachineMetadata? = nil,
        metadataVersion: Int,
        runnerState: RunnerState? = nil,
        runnerStateVersion: Int,
        health: MachineHealth? = nil
    ) {
        self.id = id
        self.namespace = namespace
        self.seq = seq
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.active = active
        self.activeAt = activeAt
        self.metadata = metadata
        self.metadataVersion = metadataVersion
        self.runnerState = runnerState
        self.runnerStateVersion = runnerStateVersion
        self.health = health
    }
}

/// A flat `machine-updated` patch.
///
/// Mirrors `MachinePatchSchema` (`shared/src/schemas.ts:509-513`), zod
/// `.strict()` — strictness discriminates the
/// `Machine | MachinePatch | null` union (see `MachineUpdatedData`).
public struct MachinePatch: Equatable, Sendable {
    public var active: Bool?
    public var activeAt: Int?
    public var updatedAt: Int?

    public init(active: Bool? = nil, activeAt: Int? = nil, updatedAt: Int? = nil) {
        self.active = active
        self.activeAt = activeAt
        self.updatedAt = updatedAt
    }
}

extension MachinePatch: Codable {
    private enum CodingKeys: String, CodingKey, CaseIterable {
        case active
        case activeAt
        case updatedAt
    }

    public init(from decoder: Decoder) throws {
        try rejectUnknownKeys(
            in: decoder,
            known: Set(CodingKeys.allCases.map(\.stringValue)),
            payloadName: "MachinePatch"
        )
        let container = try decoder.container(keyedBy: CodingKeys.self)
        active = try container.decodeIfPresent(Bool.self, forKey: .active)
        activeAt = try container.decodeIfPresent(Int.self, forKey: .activeAt)
        updatedAt = try container.decodeIfPresent(Int.self, forKey: .updatedAt)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encodeIfPresent(active, forKey: .active)
        try container.encodeIfPresent(activeAt, forKey: .activeAt)
        try container.encodeIfPresent(updatedAt, forKey: .updatedAt)
    }
}

File: android/core/protocol/src/main/kotlin/app/hapi/protocol/wire/Machine.kt
package app.hapi.protocol.wire

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

/**
 * Machine (runner host) row (`MachineSchema`, `shared/src/schemas.ts`).
 * Runner internals that v1 does not branch on stay raw [JsonElement].
 */
@Serializable
data class Machine(
    val id: String,
    val namespace: String,
    val seq: Long,
    val createdAt: Long,
    val updatedAt: Long,
    val active: Boolean,
    val activeAt: Long,
    val metadata: MachineMetadata? = null,
    val metadataVersion: Long,
    val runnerState: RunnerState? = null,
    val runnerStateVersion: Long,
    val health: MachineHealth? = null,
)

/** `MachineMetadataSchema`. */
@Serializable
data class MachineMetadata(
    val host: String,
    val platform: String,
    val happyCliVersion: String,
    val displayName: String? = null,
    val homeDir: String? = null,
    val happyHomeDir: String? = null,
    val happyLibDir: String? = null,
    val workspaceRoots: List<String>? = null,
    /** Machine-scoped RPC capability ids this runner registers. */
    val capabilities: List<String>? = null,
    @Serializable(with = LenientEpochMs::class)
    val startedCliMtimeMs: Long? = null,
    @Serializable(with = LenientEpochMs::class)
    val installedCliMtimeMs: Long? = null,
    val supervisedRestart: Boolean? = null,
)

/**
 * `RunnerStateSchema` — cheaply-typed scalars only; `capabilities` (agent
 * config descriptors) and `lastSpawnError` stay raw until a feature needs them.
 */
@Serializable
data class RunnerState(
    /** `'running' | 'shutting-down'` or any future string (zod union with string). */
    val status: String,
    val pid: Int? = null,
    val httpPort: Int? = null,
    val startedAt: Long? = null,
    val capabilities: JsonElement? = null,
    val shutdownRequestedAt: Long? = null,
    val shutdownSource: String? = null,
    val lastSpawnError: JsonElement? = null,
)

/** `MachineHealthSchema`. */
@Serializable
data class MachineHealth(
    val collectedAt: Long,
    val cpuCount: Int? = null,
    val load1m: Double? = null,
    val cpuPercent: Double? = null,
    val memoryPercent: Double? = null,
    val uptimeSeconds: Double? = null,
)

/**
 * Flat patch variant of the `machine-updated` SSE payload
 * (`MachinePatchSchema` — strict in zod).
 */
@Serializable
data class MachinePatch(
    val active: Boolean? = null,
    val activeAt: Long? = null,
    val updatedAt: Long? = null,
)

object MachinePatches {
    private val KNOWN_KEYS = setOf("active", "activeAt", "updatedAt")

    /**
     * Strict + non-empty parse mirroring the web's `getMachinePatch`
     * (`MachinePatchSchema.safeParse` + key-count check): any unknown key or
     * mistyped value → `null` (caller falls back to refetching machines), and
     * an empty `{}` is also `null`. Strictness is what separates a patch from
     * a full [Machine] payload.
     */
    fun parse(element: JsonElement?): MachinePatch? {
        val obj = element as? JsonObject ?: return null
        if (obj.isEmpty() || obj.keys.any { it !in KNOWN_KEYS }) return null
        val active = obj["active"]?.let { it.boolOrNull ?: return null }
        val activeAt = obj["activeAt"]?.let { it.longOrNull ?: return null }
        val updatedAt = obj["updatedAt"]?.let { it.longOrNull ?: return null }
        return MachinePatch(active = active, activeAt = activeAt, updatedAt = updatedAt)
    }
}
